src/analysis/normed_space/lp_space.lean

```diff
@@ -0,0 +1,519 @@
+/-
+Copyright (c) 2021 Heather Macbeth. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Heather Macbeth
+-/
+import analysis.mean_inequalities
+import analysis.mean_inequalities_pow
+import topology.algebra.ordered.liminf_limsup
+
+/-!
+# ℓp space
+
+This file describes properties of elements `f` of a pi-type `Π i, E i` with finite seminorm,
+defined for `p:ℝ≥0∞` as the size of the support of `f` if `p=0`, `(∑' a, ∥f a∥^p) ^ (1/p)` for
+`0 < p < ∞` and `⨆ a, ∥f a∥` for `p=∞`.
+
+The Prop-valued `mem_ℓp f p` states that a function `f : Π i, E i` has finite seminorm according
+to the above definition; that is, `f` has finite support if `p = 0`, `summable (λ a, ∥f a∥^p)` if
+`0 < p < ∞`, and `bdd_above (norm '' (set.range f))` if `p = ∞`.
+
+The space `Lp E p` is the subtype of elements of `Π i : α, E i` which satisfy `mem_ℓp f p`. For
+`1 ≤ p`, the seminorm is genuinely a norm and `Lp` is a complete metric space.
+
+## Main definitions
+
+* `mem_ℓp f p` : property that the function `f` satisfies, as appropriate, `f` finitely supported
+  if `p = 0`, `summable (λ a, ∥f a∥^p)` if `0 < p < ∞`, and `bdd_above (norm '' (set.range f))` if
+  `p = ∞`
+* `Lp E p` : elements of `Π i : α, E i` such that `mem_ℓp f p`. Defined as an `add_subgroup` of
+  `Π i : α, E i`.
+
+## Implementation
+
+Since `Lp` is defined as an `add_subgroup`, dot notation does not work. Use `Lp.norm_neg f` to
+say that `∥-f∥ = ∥f∥`, instead of the non-working `f.norm_neg`.
+
+## TODO
+
+* Hölder's inequality
+* Completeness of `Lp`
+* Equivalence with `pi_Lp`, for `α` finite
+* Equivalence with `measure_theory.Lp`, for `f : α → E` (i.e., functions rather than pi-types) and
+  the counting measure on `α`
+* Equivalence with `bounded_continuous_function`, for `f : α → E` (i.e., functions rather than
+  pi-types) and `p = ∞`, and the discrete topology on `α`
+
+-/
+
+noncomputable theory
+open_locale nnreal ennreal big_operators
+
+variables {α : Type*} {E : α → Type*} {p q : ℝ≥0∞} [Π i, normed_group (E i)]
+
+section ℓp
+
+/-!
+### `mem_ℓp` predicate
+
+-/
+
+section ℓp_space_definition
+
+/-- The property that `f : Π i : α, E i`
+* is finitely supported, if `p = 0`, or
+* admits an upper bound for `set.range (λ i, ∥f i∥)`, if `p = ∞`, or
+* has the series `∑' i, ∥f i∥ ^ p` be summable, if `0 < p < ∞`. -/
+def mem_ℓp (f : Π i, E i) (p : ℝ≥0∞) : Prop :=
+if p = 0 then (set.finite {i | f i ≠ 0}) else
+  (if p = ∞ then bdd_above (set.range (λ i, ∥f i∥)) else summable (λ i, ∥f i∥ ^ p.to_real))
+
+end ℓp_space_definition
+
+lemma mem_ℓp_zero {f : Π i, E i} (hf : set.finite {i | f i ≠ 0}) : mem_ℓp f 0 :=
+(if_pos rfl).mpr hf
+
+lemma mem_ℓp_infty {f : Π i, E i} (hf : bdd_above (set.range (λ i, ∥f i∥))) : mem_ℓp f ∞ :=
+(if_neg ennreal.top_ne_zero).mpr ((if_pos rfl).mpr hf)
+
+lemma mem_ℓp_gen (hp : 0 < p.to_real) {f : Π i, E i} (hf : summable (λ i, ∥f i∥ ^ p.to_real)) :
+  mem_ℓp f p :=
```

Comment:
Should `hp` be something like `p ≠ ⊤` or something like that, here and in many lemmas below? If nothing else that seems easier to use afterwards.

Reply:
I'm going to experiment with a few variants on this setup ... it's indeed a harder lemma to establish than the pair `p ≠ ⊤`, `p ≠ 0`, but it's more frequently used within the proofs than either of those facts.

Reply:
But also the lemma is true for `p=0`, no?

Reply:
Ah, indeed!

Reply:
For the other issue (the one I first thought you meant), I have thought of an approach and am trying it out on `measure_theory.Lp`, #11047.

Reply:
Actually, it ought not to be true for `p = 0`.  In that case `∥f i∥ ^ p.to_real = 1`, so it's only summable if `α` is a fintype.